src/services/WalletService.ts

```diff
@@ -133,83 +154,105 @@ export default abstract class WalletService {
-        if (addresses.length === 0) {
-          throw new Error('Error getting list of addresses - List of addresses is empty');
-        }
+      while (this.hasSubscribers() && !this.areEnoughUnusedAddresses()) {
         // eslint-disable-next-line no-await-in-loop
-        addresses = await WalletService.getUnusedValue(addresses);
-        // eslint-disable-next-line no-await-in-loop
-        const balancesFound = await ApiService.getBalances(sessionId, addresses);
-        const balances = {
-          legacy: new SatoshiBig(balancesFound.legacy || 0, 'satoshi'),
-          segwit: new SatoshiBig(balancesFound.segwit || 0, 'satoshi'),
-          nativeSegwit: new SatoshiBig(balancesFound.nativeSegwit || 0, 'satoshi'),
-        };
-        if (startFrom + maxAddressPerCall >= (this.getWalletMaxCall() * maxAddressPerCall)) {
-          this.loadingBalances = false;
-        }
-        // eslint-disable-next-line no-extra-boolean-cast
-        if (!!balances) {
-          if (balances.legacy.gt(0)
-            || balances.nativeSegwit.gt(0)
-            || balances.segwit.gt(0)) {
-            balanceAccumulated = {
-              legacy: new SatoshiBig(balanceAccumulated.legacy.plus(balances.legacy), 'satoshi'),
-              segwit: new SatoshiBig(balanceAccumulated.segwit.plus(balances.segwit), 'satoshi'),
-              nativeSegwit: new SatoshiBig(balanceAccumulated.nativeSegwit.plus(balances.nativeSegwit), 'satoshi'),
-            };
-          } else {
-            const areAllAddressUnused = addresses
-              .every((walletAddressItem) => walletAddressItem.unused);
-            if (areAllAddressUnused) return;
-          }
-          this.informSubscribers(balanceAccumulated, addresses);
-        } else {
-          throw new Error('Error getting balances');
-        }
+        await this.askForBalance(sessionId);
         const maxAmountPeginCompare = new SatoshiBig(maxAmountPegin, 'satoshi');
-        if (balanceAccumulated.legacy.gte(maxAmountPeginCompare)
-          && balanceAccumulated.segwit.gte(maxAmountPeginCompare)
-          && balanceAccumulated.nativeSegwit.gte(maxAmountPeginCompare)
+        if (this.balanceAccumulated.legacy.gte(maxAmountPeginCompare)
+          && this.balanceAccumulated.segwit.gte(maxAmountPeginCompare)
+          && this.balanceAccumulated.nativeSegwit.gte(maxAmountPeginCompare)
         ) {
           return;
         }
+        this.setAddressesToFetch();
+        const maxIndexReached = Math.max(
+          this.addressesToFetch.legacy.lastIndex,
+          this.addressesToFetch.segwit.lastIndex,
+          this.addressesToFetch.nativeSegwit.lastIndex,
+        );
+        if (maxIndexReached >= maxAddressesHardStop) {
+          return;
+        }
       }
     } catch (error: any) {
       throw new Error(`Balance Error: ${error.message}`);
     } finally {
       this.loadingBalances = false;
+      this.informSubscribers(this.balanceAccumulated, []);
```

Comment:
Can you change this line to just before the catch statement? In the tests I did, even in case of an error, the user is still redirected to the next screen.